base_layer/core/src/transaction_protocol_manager.rs

```diff
@@ -0,0 +1,1046 @@
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE
+
+//! The Transaction Protocol Manager implements a protocol to construct a Mimwblewimble transaction between two parties
+//! , a Sender and a Receiver. In this transaction the Sender is paying the Receiver from their inputs and also paying
+//! to as many change outputs as they like. The Receiver will receive a single output from this transaction.
+//! The module is written to allow for a single entity to construct a transaction by moving through the state machine
+//! and providing all the required data OR for two remote parties to collaborate to build the transaction by having the
+//! Transaction Protocol Manager construct and accept messages with the public information required from the two
+//! parties. Wallets are an example of an application that will often be constructing transactions to pay themselves as
+//! part of UTXO management. Peer-to-Peer payments will make use of the public messages for the remote parties to
+//! collaborate.
+//!
+//! Below is an illustration of the single user state machine flow that this module implements, the transitions in the
+//! diagram represent the various events that the state machine can handle.
+//!
+//! +------------+
+//! |            +-------+
+//! |   Sender   |       | SetFee
+//! |   Setup    +<------+
+//! |            |
+//! |            +-------+
+//! |            |       | SetLockHeight
+//! |            +<------+
+//! |            |
+//! |            +-------+
+//! |            |       | SetAmount
+//! |            +<------+
+//! |            |
+//! |            +-------+
+//! |            |       | AddInput                                   +---------+
+//! |            +<------+                                            | \o/ Fin |
+//! |            |                                                    +----+----+
+//! |            +-------+                                                 ^
+//! |            |       | AddChangeOutput                                 |
+//! |            +<------+                                                 |
+//! |            |                                                   +-----+------+
+//! |            +-------+                                           |            |
+//! |            |       | SetSenderNonce                            | Finalized  |
+//! |            +<------+                                           | Transaction|
+//! |            |                                                   |            |
```

Comment:
Seems like these are more builder methods that could create the initial state. Can they be run in any order?

Reply:
Correct, they can be run in any order and do not leave that state so in that sense they are builder style methods.

Reply:
So after thinking about it bit more I do agree, I think I am going to make the initial setup (currently the SenderSetup state) into a builder and when it can be built starts the state machine. The Receiver will start the state machine by ingesting the public message from the sender. This will reduce the number of events significantly.